File: java/leetcode/src/main/java/com/leetcode/list/Add_Two_Numbers_2.java
package com.leetcode.list;

public class Add_Two_Numbers_2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return add(l1, l2, 0);
    }

    ListNode add(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 ==  null) {
            if (carry == 1) return new ListNode(1);
            return null;
        }

        int sum = carry;
        sum = l1 == null ? sum : sum + l1.val;
        sum = l2 == null ? sum : sum + l2.val;

        int num = sum % 10;
        carry = sum / 10;

        ListNode node = new ListNode(num);
        if (l1 == null) {
            node.next = add(null, l2.next, carry);
        } else if(l2 == null) {
            node.next = add(l1.next, null, carry);
        } else {
            node.next = add(l1.next, l2.next, carry);
        }
        return node;
    }
}

File: js/algrithm/huahua130/list/Add_Two_Numbers_2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    const add = function(l1, l2, carry) {
        if (!l1 && !l2 && !carry) return null
        let sum = carry
        sum += l1 ? l1.val : 0
        sum += l2? l2.val : 0

        let nextNode = undefined
        if (l1 && l2) {
            nextNode = add(l1.next, l2.next, Math.floor(sum / 10))
        } else if (l1) {
            nextNode = add(l1.next, null, Math.floor(sum / 10))
        } else if (l2) {
            nextNode = add(null, l2.next, Math.floor(sum / 10))
        } else {
            nextNode = undefined
        }
        return new ListNode(sum % 10, nextNode)
    }

    return add(l1, l2, 0)
};